Show deleted files finalize

kevinrobbins opened the pull request.

djkiourtsis committed:
Add method to get deleted s3 contents

This adds the getDeletedContents method to the AmazonS3Backend.  This
will collect and return the delete markers for an s3 folder.

djkiourtsis committed:
Add deleted folders and files to bucket scripts

This updates the bucket controller and html to allow for the viewing
of deleted s3 files and folders.  This can be enabled through a
checkbox when using s3commander.

kevinrobbins committed:
Remove delete button for deleted files

Deleted files cannot be deleted again, so only show the Delete button
if the file exists.

kevinrobbins committed:
Remove delete button for deleted folders

Deleted folders cannot be deleted again so only show the Delete button
for existing folders.

kevinrobbins committed:
Change background color for deleted files and folders

The only way to tell the difference between deleted and non-deleted
files/folders was if a Delete button was present. Now the background
color is red if the file/folder is deleted (shown by checking the "Show
deleted files" checkbox).

kevinrobbins committed:
Remove folders that aren't deleted when "Show deleted files" is checked

I put the code that filters out the non-deleted folders in
loadContents() because it relies on getContents() to have already been
run and completed to produce accurate results. If getContents() hasn't
been run, this.folders could have stale contents or no contents.

I could see an argument for putting this into the
getDeletedContents method because the method name implies it
returns only deleted content, but it returns all folders regardless
of whether or not they are deleted which is misleading.

If performance isn't an issue, this logic could be moved to
getDeletedContents() if getDeletedContents() calls s3.getObjectsV2()
itself. In this case getDeletedContents() would no longer rely on
loadContents(), but the two separate calls to s3.getObjectsV2() could
return different results if a change to the contents of the current path
happens in between. Maybe that's not a big deal? (It occurs to me as I
write this that this problem could also occur if a change occurs between
s3.getObjectsV2() and s3.listObjectVersions())

In the end, I decided that having the method produce consistent
results was worth the trade off of requiring additonal post-processing.

kevinrobbins committed:
Initialize fields on BucketController class

All other fields were initialized in the constructor of the
BucketController except deletedFolders and deletedFiles. Adding them
for uniformity.

kevinrobbins committed:
Only display the latest version of a file when viewing deleted files

Prior to this change, if "Show deleted files" was checked then any file
that had a deleteMarker in its history would be displayed even if the
most recent version of the file did not have a delete marker. This
scenario happens when a file is deleted and then a new file with the
same name is created in the same directory as the original file.

Checking the isLatest flag ensures that the file will only be displayed
if the most recent version has a delete marker.

kevinrobbins committed:
Make entire "Show deleted files" checkbox clickable

Only the checkbox was clickable. This makes it easier to check the box.

kevinrobbins committed:
Update view with the current directory's deleted files

The deleted files were not being refreshed correctly when navigating
throught the tree. The previously viewed directory's deleted files were
shown instead of the currently viewed directory's deleted files.

kevinrobbins committed:
Modify code to resolve console warnings

kevinrobbins commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Implement a toggle in S3Commander to show deleted files and folders.](https://trello.com/c/zzXEGXir/2503-implement-a-toggle-in-s3commander-to-show-deleted-files-and-folders)

CtrlC-Root commented on src/bucket/BucketController.ts:
```diff
@@ -167,6 +183,32 @@ export class BucketController {
         this.files = contents.files.sort(compareObjectNames);
+
+        this.backend.getDeletedContents(this.bucket, this.currentFolder)
+        .then((contents: IFolderContents) => {
```
This formatting is not consistent with the rest of the project. The `.then` and `.catch` lines need to be indented one level past the top-level `this.backend` line.

CtrlC-Root commented on src/file/file.html:
```diff
@@ -1,8 +1,8 @@
   <a href="{{ $ctrl.file.getDownloadLink() }}"
     download="{{ $ctrl.file.getPath().name() }}"
-    target="_blank">{{ $ctrl.file.getPath().name() }}</a>
+    target="_blank" >{{ $ctrl.file.getPath().name() }}</a>
```
Unnecessary whitespace.

CtrlC-Root commented:
Overall the code looks reasonable. I would have modified the `getContents` method or added a new one called `getVersionedContents` which could be used to retrieve all available versions of each file and folder. Maybe we'll come back and revisit this later before making the next major release.

ericamador commented on src/index.css:
```diff
@@ -195,3 +195,7 @@ div.s3c-controls label {
+
+div.deleted {
+  background-color: #ffcccc;
+}
```
Needs newline at end of file. Fix this in a rebase - don't add a new commit for small formatting issues like this.

kevinrobbins replied:
I have already fixed this, I just haven't pushed it up yet.

kevinrobbins committed:
Change deleted file names to spans from links

Deleted files cannot be downloaded. Originally, deleted file names were
displayed as anchors blank href attributes. This produced odd behavior
in Django CMS (not sure if it actually affected live sites or not, but
disabling the CMS toolbar did not resolve the issue). The behavior
observed was the link was clickable and it downloaded the current page's
source code (as if the user had right-clicked and selected View Source).

While not a big deal even if this reached the live site, it was worth
changing to the deleted file names to be spans since they don't have
download links anyway.

kevinrobbins committed:
Change deleted file names color from black to dark red

When viewing in a tss site, the non-deleted files have green text set
through unify. To better match the theme on the tss site, I changed the
name color to darkred.
